src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -133,38 +133,129 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
+                {
+                    argumentRefKindsOpt = argumentRefKindsOpt.Insert(0, thisRefKind);
+                }
+
+                // PROTOTYPE(ic): search for a scenario where propagating this value matters
+                invokedAsExtensionMethod = true;
+            }
+
+            method = interceptor;
+
+            return;
+        }
+
         public override BoundNode VisitCall(BoundCall node)
         {
             Debug.Assert(node != null);
 
+            MethodSymbol method = node.Method;
+            BoundExpression? receiverOpt = node.ReceiverOpt;
+            ImmutableArray<BoundExpression> arguments = node.Arguments;
+            ImmutableArray<int> argsToParamsOpt = node.ArgsToParamsOpt;
+            ImmutableArray<RefKind> argRefKindsOpt = node.ArgumentRefKindsOpt;
+            bool invokedAsExtensionMethod = node.InvokedAsExtensionMethod;
+
             // Rewrite the receiver
-            BoundExpression? rewrittenReceiver = VisitExpression(node.ReceiverOpt);
-            var argRefKindsOpt = node.ArgumentRefKindsOpt;
+            BoundExpression? rewrittenReceiver = VisitExpression(receiverOpt);
 
             ArrayBuilder<LocalSymbol>? temps = null;
             var rewrittenArguments = VisitArgumentsAndCaptureReceiverIfNeeded(
                 ref rewrittenReceiver,
                 captureReceiverMode: ReceiverCaptureMode.Default,
-                node.Arguments,
-                node.Method,
-                node.ArgsToParamsOpt,
+                arguments,
+                method,
+                argsToParamsOpt,
                 argRefKindsOpt,
                 storesOpt: null,
                 ref temps);
 
-            var rewrittenCall = MakeArgumentsAndCall(
-                syntax: node.Syntax,
-                rewrittenReceiver: rewrittenReceiver,
-                method: node.Method,
-                arguments: rewrittenArguments,
-                argumentRefKindsOpt: argRefKindsOpt,
-                expanded: node.Expanded,
-                invokedAsExtensionMethod: node.InvokedAsExtensionMethod,
-                argsToParamsOpt: node.ArgsToParamsOpt,
-                resultKind: node.ResultKind,
-                type: node.Type,
-                temps,
-                nodeOpt: node);
+            rewrittenArguments = MakeArguments(
+                node.Syntax,
+                rewrittenArguments,
+                method,
+                node.Expanded, // PROTOTYPE(ic): params differences shouldn't matter--maybe even make it an error--but we need to test
+                argsToParamsOpt,
+                ref argRefKindsOpt,
+                ref temps,
+                invokedAsExtensionMethod);
+
+            InterceptCallAndAdjustArguments(ref method, ref rewrittenReceiver, ref rewrittenArguments, ref argRefKindsOpt, ref invokedAsExtensionMethod, node.InterceptableLocation);
+
+            // PROTOTYPE(ic): intercept with object.ReferenceEquals?
```

Comment:
>intercept with object.ReferenceEquals? [](http://example.com/codeflow?start=30&length=38)

What does this comment mean? Is this a specific test that should be added?

Reply:
It appears that calling `object.ReferenceEquals` is specially defined to do the same thing as the built-in `==` operator on `object`. So the question would be, if the interceptor *is* `object.ReferenceEquals`, do we expect that behavior to occur.

Reply:
https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYVwG9NcL88EAWXAWQAoB7YAKwFMBjAF11bYoANP3bc+ApAEpS5SvKgAzXCzG8AdACUOijgCcOMLhwCiARwCuEADYBnVYJGSpMkkWYxbzax3UB1PSgeDkYUKQBuXABfOQoYjCigA=

Reply:
Are there other methods for which the compiler has special knowledge (maybe during emitting or nullability analysis) and that we'd want to prevent from being intercepted? For example `GetValueOrDefault` (in `CodeGenerator.IsReadOnlyCall`)

Reply:
Likely yes. I feel "neutral" about whether actually intercepting such special calls should be permitted. I would just want the behavior to meet some "experimental" quality bar: compiler doesn't crash, generated code doesn't crash, and perhaps if we decide not to intercept we give an error.